src/main/java/org/apache/mesos/scheduler/repair/RepairScheduler.java

```diff
@@ -0,0 +1,163 @@
+
+        Optional<OfferRequirement> offerReq = Optional.empty();
+        boolean replacingStopped = false;
+
+        if (terminatedTasks.size() > 0) {
+            TaskInfo terminatedTask = terminatedTasks.get(new Random().nextInt(terminatedTasks.size()));
+            log.info("Found terminated task");
+            if (failureMonitor.hasFailed(terminatedTask)) {
+                log.info("Purging terminated task");
+                failureListener.taskFailed(terminatedTask.getTaskId());
+            } else {
+                log.info("Relaunching terminated task in place");
+                offerReq = Optional.of(offerReqProvider.getReplacementOfferRequirement(terminatedTask));
+            }
+        } else {
+            log.info("Launching a stopped task");
+            replacingStopped = true;
+            offerReq = offerReqProvider.maybeGetNewOfferRequirement(targetConfigName, block);
+        }
+
+        if (offerReq.isPresent() && launchConstrainer.canLaunch(offerReq.get())) {
+            log.info("Preparing to launch task");
+            OfferEvaluator offerEvaluator = new OfferEvaluator();
+            List<OfferRecommendation> recommendations = offerEvaluator.evaluate(offerReq.get(), offers);
+            if (!recommendations.isEmpty()) {
+                // we're replacing some task(s)
+                Set<String> relaunchedTasks = recommendations.stream()
+                        .filter(it -> it.getOperation().hasLaunch())
+                        .flatMap(it -> it.getOperation().getLaunch().getTaskInfosList().stream())
+                        .map(TaskInfo::getName)
+                        .collect(Collectors.toSet());
+                if (replacingStopped) {
+                    // we'll filter their names out of the stopped pool
+                    newStopped = newStopped.stream()
+                            .filter(it -> !relaunchedTasks.contains(it))
+                            .collect(Collectors.toList());
+                } else {
+                    // we'll filter their names out of the failed pool
+                    newFailed = newFailed.stream()
+                            .filter(it -> !relaunchedTasks.contains(it))
+                            .collect(Collectors.toList());
+                }
+            }
+            List<Operation> launchOperations = recommendations.stream()
+                    .map(OfferRecommendation::getOperation)
+                    .filter(Operation::hasLaunch)
+                    .collect(Collectors.toList());
+            if (launchOperations.size() > 1) {
+                throw new RuntimeException("Repairs should only launch one task at a time, found "
+                        + launchOperations.size() + " tasks");
+            }
+            acceptedOffers = offerAccepter.accept(driver, recommendations);
+            if (launchOperations.size() == 1) {
+                // We could've launched nothing if the offer didn't fit; only make if it happened
```

Comment:
I deleted what comes after the semicolon; it was meant to emphasize we only make the invocation to `launchHappened` if the launch actually happened, rather than nearly happened (since we unconditionally accept recommendations.